### DISCUSSION TITLE
How to block client creation on etcd being up?

### DISCUSSION THREAD

**Discussion by mayooot** [etcd-io/etcd#17197] [Q&A] [answered]
### What would you like to be added?

When my local etcd is not started, I run the following code. What confuses me is that err is nil.
~~~go
package main

import (
	"time"

	clientv3 "go.etcd.io/etcd/client/v3"
)

func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   []string{"127.0.0.1:2379"},
		DialTimeout: 1 * time.Second,
	})
	if err != nil {
		panic(err.Error())
	}
	defer cli.Close()
}

~~~

Without changing the `clientv3.New()` method, provide a ping method that returns an error if etcd does not start.
Like this. [commit](https://github.com/mayooot/etcd/commit/3e898df0d504fb097e3328eb1c6f20c66d2ccf5d)

It is used as follows:
~~~go
var err error
cli, err = clientv3.New(clientv3.Config{
    Endpoints:   []string{"127.0.0.1:2379"},
    DialTimeout: 1 * time.Second,
})
if err != nil {
    panic(err.Error())
}
defer cli.Close()

ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
defer cancel()
if err := cli.Ping(ctx); err != nil {
    panic(err)
}

~~~

### Why is this needed?

Because I didn't know if the client's state was successful or failed after I called `clientv3.New()` to initialize it, the program hangs during a put operation.

**Comment by mayooot** [comment]
My Development Environment:
~~~
$ sw_vers
ProductName:		macOS
ProductVersion:		14.0
BuildVersion:		23A344

$ sysctl -n machdep.cpu.brand_string
Apple M1

$ go version
go version go1.21.5 darwin/arm64
~~~

**Comment by serathius** [comment, marked as answer]
By default grpc client will not block on startup. You can set `DialOptions: []grpc.DialOption{grpc.WithBlock()},` to get the behavior you want.

**Comment by mayooot** [comment]
> By default grpc client will not block on startup. You can set `DialOptions: []grpc.DialOption{grpc.WithBlock()},` to get the behavior you want.

Thanks, you are correct, if `Endpoints` don't exist, the program will keep blocking. But I didn't find out how to set the timeout when initializing the `clientv3.Client`, so I used the follwing.
~~~go
package main

import (
	"context"
	"time"

	clientv3 "go.etcd.io/etcd/client/v3"
	"google.golang.org/grpc"
)

var cli *clientv3.Client

func main() {
	if err := Init(); err != nil {
		panic(err)
	}
}

func Init() error {
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()

	resultCh := make(chan *clientv3.Client, 1)
	errCh := make(chan error, 1)
	go func() {
		cli, err := clientv3.New(clientv3.Config{
			Endpoints:   []string{"127.0.0.1:2379"},
			DialOptions: []grpc.DialOption{grpc.WithBlock()},
		})
		if err != nil {
			errCh <- err
			return
		}
		resultCh <- cli
	}()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case err := <-errCh:
		return err
	case cli = <-resultCh:
		return nil
	}
}

~~~
Do you have any better practice for initializing `clientv3.Client`?

**Comment by serathius** [comment]
DialTimeout doesn't work?

**Comment by mayooot** [comment]
> DialTimeout doesn't work?

sorry, I forgot to set `DialTimeout`.
Thanks a Ton. ☺️
